bloomfilter/src/Data/BloomFilter.hs

```diff
@@ -0,0 +1,265 @@
+{-# LANGUAGE BangPatterns             #-}
+{-# LANGUAGE Rank2Types               #-}
+{-# LANGUAGE RoleAnnotations          #-}
+{-# LANGUAGE ScopedTypeVariables      #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeApplications         #-}
+{-# LANGUAGE TypeOperators            #-}
+-- |
+--
+-- A fast, space efficient Bloom filter implementation.  A Bloom
+-- filter is a set-like data structure that provides a probabilistic
+-- membership test.
+--
+-- * Queries do not give false negatives.  When an element is added to
+--   a filter, a subsequent membership test will definitely return
+--   'True'.
+--
+-- * False positives /are/ possible.  If an element has not been added
+--   to a filter, a membership test /may/ nevertheless indicate that
+--   the element is present.
+--
+-- This module provides low-level control.  For an easier to use
+-- interface, see the "Data.BloomFilter.Easy" module.
+
+module Data.BloomFilter (
+    -- * Overview
+    -- $overview
+
+    -- ** Ease of use
+    -- $ease
+
+    -- ** Performance
+    -- $performance
+
+    -- ** Differences from bloomfilter package
+    -- $differences
+
+    -- * Types
+    Hash,
+    Bloom,
+    MBloom,
+    Bloom',
+    MBloom',
+    CheapHashes,
+    RealHashes,
+
+    -- * Immutable Bloom filters
+
+    -- ** Conversion
+    freeze,
+    thaw,
+    unsafeFreeze,
+
+    -- ** Creation
+    unfold,
+
+    fromList,
+    empty,
+    singleton,
+
+    -- ** Accessors
+    length,
+    elem,
+    elemHashes,
+    notElem,
+) where
+
+import           Control.Monad (forM_, liftM)
+import           Control.Monad.ST (ST, runST)
+import           Data.BloomFilter.Hash (CheapHashes, Hash, Hashable,
+                     Hashes (..), RealHashes)
+import           Data.BloomFilter.Internal (Bloom' (..))
+import           Data.BloomFilter.Mutable (MBloom, MBloom', insert, new)
+import qualified Data.BloomFilter.Mutable.Internal as MB
+import           Data.Word (Word64)
+
+import           Prelude hiding (elem, length, notElem)
+
+import qualified Data.BloomFilter.BitVec64 as V
+
+-- | Bloom filter using 'CheapHashes' hashing scheme.
+type Bloom = Bloom' CheapHashes
+
+-- | Create an immutable Bloom filter, using the given setup function
+-- which executes in the 'ST' monad.
+--
+-- Example:
+--
+-- @
+-- TODO
+--import "Data.BloomFilter.Hash" (cheapHashes)
+--
+--filter = create (cheapHashes 3) 1024 $ \mf -> do
+--           insertMB mf \"foo\"
+--           insertMB mf \"bar\"
+-- @
+--
+-- Note that the result of the setup function is not used.
+create :: Int        -- ^ number of hash functions to use
```

Comment:
This is non-exported function, thus I didn't cleanup the haddock example (which is very dated)